File: 0127-word-ladder/0127-word-ladder.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        if endWord not in wordList:
            return 0
        if beginWord not in wordList:
            wordList.append(beginWord)

        hashMap = dict()

        for i in range(len(wordList)):
            w = wordList[i]

            for j in range(len(w)):
                word = w[0:j] + "*" + w[j+ 1 : len(w)]
                # print(word)
                if word not in hashMap:
                    hashMap[word] =  list()
                hashMap[word].append(w)

        # print(hashMap)        
        queue = deque()
        queue.append(beginWord)

        visited = dict()
        visited[beginWord] = True
        level = 0
        while queue:
            size = len(queue)
            level += 1

            for i in range(size):
                word = queue.popleft()
                if word == endWord:
                    return level
                
                for j in range(len(word)):
                    newWord = word[0:j] + "*" + word[j+ 1: len(word)]
                    for adjWord in hashMap.get(newWord, []):
                        if adjWord not in visited.keys():
                            queue.append(adjWord) 
                            visited[adjWord]= True
        
        return 0



         

File: 0127-word-ladder/0127-word-ladder.java
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        if(beginWord == null || endWord == null){
            return 0;
        }        
        if(beginWord == endWord){
            return 0;
        }
        boolean exist = false;
        for(int i = 0; i< wordList.size(); i++){
            if(beginWord.equals(wordList.get(i))){
                exist = true;
                break;
            }
        }
        if(exist == false){
            wordList.add(beginWord);
        }
        HashMap<String, List<String>> map = new HashMap<>();
        
        for(int i = 0; i< wordList.size(); i++){
            String word = wordList.get(i);

            for(int j = 0; j < word.length(); j++){
                String newWord = word.substring(0, j) + '*' + word.substring(j + 1, word.length());
                if(!map.containsKey(newWord)){
                    map.put(newWord, new ArrayList<>());
                }
                map.get(newWord).add(word);
            }
        }

        Queue<String> q = new LinkedList<>();
        int level = 0;
        q.add(beginWord);
        // level++;
        HashMap<String, Boolean> visited = new HashMap<>();
        visited.put(beginWord, true);

        while(!q.isEmpty()){
            int size = q.size();
            level++;
            for(int i = 0; i< size; i++){
                String word = q.poll();

                if (word.equals(endWord)){
                    return level;
                }
                for(int j = 0; j < word.length(); j++){
                    String newWord = word.substring(0, j) +  '*' + word.substring(j + 1, word.length());
                    List<String> li = map.getOrDefault(newWord, new ArrayList<>());
                    for(String adjWord: li){
                        if(!visited.containsKey(adjWord)){
                            visited.put(adjWord,true);
                            q.add(adjWord);
                        }
                    }
                }
            }
        }
                        return 0;
    }
}